// @fortawesome/angular-fontawesome/esm5/layers/layers-counter.component.js
import { __decorate, __param } from "tslib";
import { Component, HostBinding, Input, Optional } from '@angular/core';
import { DomSanitizer, SafeHtml } from '@angular/platform-browser';
import { counter } from '@fortawesome/fontawesome-svg-core';
import { faWarnIfParentNotExist } from '../shared/errors/warn-if-parent-not-exist';
import { FaLayersComponent } from './layers.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/platform-browser';
var FaLayersCounterComponent = /** @class */ (function () {
    function FaLayersCounterComponent(parent, sanitizer) {
        this.parent = parent;
        this.sanitizer = sanitizer;
        this.classes = [];
        faWarnIfParentNotExist(this.parent, 'FaLayersComponent', this.constructor.name);
    }
    FaLayersCounterComponent.prototype.ngOnChanges = function (changes) {
        if (changes) {
            var params = this.buildParams();
            this.updateContent(params);
        }
    };
    FaLayersCounterComponent.prototype.buildParams = function () {
        return {
            title: this.title,
            classes: this.classes,
            styles: this.styles,
        };
    };
    FaLayersCounterComponent.prototype.updateContent = function (params) {
        this.renderedHTML = this.sanitizer.bypassSecurityTrustHtml(counter(this.content || '', params).html.join(''));
    };
    FaLayersCounterComponent.ctorParameters = function () { return [
        { type: FaLayersComponent, decorators: [{ type: Optional }] },
        { type: DomSanitizer }
    ]; };
    __decorate([
        Input()
    ], FaLayersCounterComponent.prototype, "content", void 0);
    __decorate([
        Input()
    ], FaLayersCounterComponent.prototype, "title", void 0);
    __decorate([
        Input()
    ], FaLayersCounterComponent.prototype, "styles", void 0);
    __decorate([
        Input()
    ], FaLayersCounterComponent.prototype, "classes", void 0);
    __decorate([
        HostBinding('innerHTML')
    ], FaLayersCounterComponent.prototype, "renderedHTML", void 0);
    FaLayersCounterComponent = __decorate([ __param(0, Optional())
    ], FaLayersCounterComponent);
FaLayersCounterComponent.ɵfac = function FaLayersCounterComponent_Factory(t) { return new (t || FaLayersCounterComponent)(ɵngcc0.ɵɵdirectiveInject(FaLayersComponent, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DomSanitizer)); };
FaLayersCounterComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FaLayersCounterComponent, selectors: [["fa-layers-counter"]], hostAttrs: [1, "ng-fa-layers-counter"], hostVars: 1, hostBindings: function FaLayersCounterComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("innerHTML", ctx.renderedHTML, ɵngcc0.ɵɵsanitizeHtml);
    } }, inputs: { classes: "classes", content: "content", title: "title", styles: "styles" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 0, vars: 0, template: function FaLayersCounterComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FaLayersCounterComponent, [{
        type: Component,
        args: [{
                selector: 'fa-layers-counter',
                template: '',
                host: {
                    class: 'ng-fa-layers-counter'
                }
            }]
    }], function () { return [{ type: FaLayersComponent, decorators: [{
                type: Optional
            }] }, { type: ɵngcc1.DomSanitizer }]; }, { classes: [{
            type: Input
        }], renderedHTML: [{
            type: HostBinding,
            args: ['innerHTML']
        }], content: [{
            type: Input
        }], title: [{
            type: Input
        }], styles: [{
            type: Input
        }] }); })();
    return FaLayersCounterComponent;
}());
export { FaLayersCounterComponent };
