// ngx-page-scroll-core/esm2015/lib/ngx-page-scroll-core.module.js
import { NgModule } from '@angular/core';
import { PageScrollService } from './providers/ngx-page-scroll.service';
import { NGXPS_CONFIG } from './providers/config.provider';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = {};
export class NgxPageScrollCoreModule {
    static forRoot(config) {
        return {
            ngModule: NgxPageScrollCoreModule,
            providers: [PageScrollService, { provide: NGXPS_CONFIG, useValue: config }],
        };
    }
}
NgxPageScrollCoreModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxPageScrollCoreModule });
NgxPageScrollCoreModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxPageScrollCoreModule_Factory(t) { return new (t || NgxPageScrollCoreModule)(); }, providers: [
        PageScrollService,
        { provide: NGXPS_CONFIG, useValue: ɵ0 },
    ] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxPageScrollCoreModule, [{
        type: NgModule,
        args: [{
                providers: [
                    PageScrollService,
                    { provide: NGXPS_CONFIG, useValue: ɵ0 },
                ]
            }]
    }], null, null); })();
export { ɵ0 };
