// @sweetalert2/ngx-sweetalert2/esm5/lib/sweetalert2-loader.service.js
import { __decorate, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { swalProviderToken } from './di';
import * as ɵngcc0 from '@angular/core';
var SweetAlert2LoaderService = /** @class */ (function () {
    // Using any because Angular metadata generator does not understand a pure TS type here
    function SweetAlert2LoaderService(swalProvider) {
        this.swalProvider = swalProvider;
    }
    Object.defineProperty(SweetAlert2LoaderService.prototype, "swal", {
        get: function () {
            if (!this.swalPromiseCache) {
                this.preloadSweetAlertLibrary();
            }
            return this.swalPromiseCache;
        },
        enumerable: true,
        configurable: true
    });
    SweetAlert2LoaderService.prototype.preloadSweetAlertLibrary = function () {
        if (this.swalPromiseCache)
            return;
        var libPromise = isLoader(this.swalProvider)
            ? this.swalProvider()
            : Promise.resolve(this.swalProvider);
        this.swalPromiseCache = libPromise.then(function (value) { return isDefaultExport(value) ? value : value.default; });
        function isLoader(value) {
            return typeof value === 'function' && value.version === undefined;
        }
        function isDefaultExport(value) {
            return typeof value === 'function';
        }
    };
    SweetAlert2LoaderService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [swalProviderToken,] }] }
    ]; };
    SweetAlert2LoaderService = __decorate([ __param(0, Inject(swalProviderToken))
    ], SweetAlert2LoaderService);
SweetAlert2LoaderService.ɵfac = function SweetAlert2LoaderService_Factory(t) { return new (t || SweetAlert2LoaderService)(ɵngcc0.ɵɵinject(swalProviderToken)); };
SweetAlert2LoaderService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: SweetAlert2LoaderService, factory: function (t) { return SweetAlert2LoaderService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SweetAlert2LoaderService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [swalProviderToken]
            }] }]; }, null); })();
    return SweetAlert2LoaderService;
}());
export { SweetAlert2LoaderService };
